pkg/reloader/reloader.go

```diff
@@ -0,0 +1,170 @@
+			reloaderLogger.Debugf("Secret %s not found in secretVersions map, creating it", secretPath)
+			newSecretVersions[secretPath] = currentVersion
+			continue
+		}
+		if c.secretVersions[secretPath] == currentVersion {
+			reloaderLogger.Debugf("Secret %s did not change", secretPath)
+			newSecretVersions[secretPath] = currentVersion
+			continue
+		}
+		reloaderLogger.Debugf("Secret version stored: %d current: %d", c.secretVersions[secretPath], currentVersion)
+		for _, workload := range workloads {
+			workloadsToReload[workload] = true
+		}
+		newSecretVersions[secretPath] = currentVersion
+	}
+
+	// Reloading workloads
+	for workload := range workloadsToReload {
+		reloaderLogger.Info("Reloading workload: ", workload)
+		err := c.reloadWorkload(workload)
+		if err != nil {
+			reloaderLogger.Error("failed reloading workload: ", workload, err)
+		}
+	}
+
+	// Replace secretVersions map with the new one so we don't keep deleted secrets in the map
+	c.secretVersions = newSecretVersions
+	reloaderLogger.Debug("Updated secretVersions map: ", newSecretVersions)
+
+	if len(workloadsToReload) == 0 {
+		reloaderLogger.Info("No workloads to reload")
+	}
+}
+
+func (c *Controller) reloadWorkload(workload workload) error {
+	// Reload object based on its type
+	switch workload.kind {
+	case DeploymentKind:
+		deployment, err := c.kubeClient.AppsV1().Deployments(workload.namespace).Get(context.Background(), workload.name, metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+
+		deployment.Spec.Template.GetAnnotations()[ReloadCountAnnotationName] = getReloadCountAnnotation(&deployment.Spec.Template)
+
+		_, err = c.kubeClient.AppsV1().Deployments(workload.namespace).Update(context.Background(), deployment, metav1.UpdateOptions{})
+		if err != nil {
+			return err
+		}
+
+	case DaemonSetKind:
+		daemonSet, err := c.kubeClient.AppsV1().DaemonSets(workload.namespace).Get(context.Background(), workload.name, metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+
+		daemonSet.Spec.Template.GetAnnotations()[ReloadCountAnnotationName] = getReloadCountAnnotation(&daemonSet.Spec.Template)
+
+		_, err = c.kubeClient.AppsV1().DaemonSets(workload.namespace).Update(context.Background(), daemonSet, metav1.UpdateOptions{})
+		if err != nil {
+			return err
+		}
+
+	case StatefulSetKind:
+		statefulSet, err := c.kubeClient.AppsV1().StatefulSets(workload.namespace).Get(context.Background(), workload.name, metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+
+		statefulSet.Spec.Template.GetAnnotations()[ReloadCountAnnotationName] = getReloadCountAnnotation(&statefulSet.Spec.Template)
+
+		_, err = c.kubeClient.AppsV1().StatefulSets(workload.namespace).Update(context.Background(), statefulSet, metav1.UpdateOptions{})
+		if err != nil {
+			return err
+		}
+
+	default:
+		return fmt.Errorf("unknown object type: %s", workload.kind)
+	}
+
+	return nil
+}
+
+func getReloadCountAnnotation(podTemplate *corev1.PodTemplateSpec) string {
```

Comment:
I think it would be better if this method updated the annotation by default rather than returning the version and then performing an update, ie. `func incrementReloadCountAnnotation(podTemplate *corev1.PodTemplateSpec) { }`

Reply:
Why? It adds a lot of complexity to that function, it would also have a switch case depending on the workload kind, it would need a k8sClient etc... I like the word `increment` though!

Reply:
There is no need to pass any extra arguments. This method will perform API requests, it will just increment the annotation value, for example 
```golang
func incrementReloadCountAnnotation(podTemplate *corev1.PodTemplateSpec) {
	var version string

	reloadCount := podTemplate.GetAnnotations()[ReloadCountAnnotationName]
	if reloadCount == "" {
		version = "1"
	} else {
		count, err := strconv.Atoi(reloadCount)
		if err != nil {
			version = "1"
		}
		count++
		version = strconv.Itoa(count)
	}

	podTemplate.GetAnnotations()[ReloadCountAnnotationName] = version
}
```

And then, in the `reloadWorkload` function, you can rely only on calling the method rather than calling and assigning
```go
incrementReloadCountAnnotation(&deployment.Spec.Template) // deployment
incrementReloadCountAnnotation(&daemonSet.Spec.Template) // daemonSet
incrementReloadCountAnnotation(&statefulSet.Spec.Template) // statefulSet
```